packages/block-library/src/query/edit/query-toolbar.js

```diff
@@ -1,30 +1,51 @@
 /**
  * WordPress dependencies
  */
-import { ToolbarGroup, ToolbarButton } from '@wordpress/components';
+import {
+	ToolbarGroup,
+	ToolbarButton,
+	Dropdown,
+	__experimentalDropdownContentWrapper as DropdownContentWrapper,
+} from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
-import { usePatterns } from '../utils';
+import PatternSelection, { useBlockPatterns } from './pattern-selection';
 
-export default function QueryToolbar( {
-	openPatternSelectionModal,
-	name,
-	clientId,
-} ) {
-	const hasPatterns = !! usePatterns( clientId, name ).length;
+export default function QueryToolbar( { clientId, attributes } ) {
+	const hasPatterns = useBlockPatterns( clientId, attributes ).length;
+	if ( ! hasPatterns ) {
+		return null;
+	}
 
 	return (
-		<>
-			{ hasPatterns && (
-				<ToolbarGroup className="wp-block-template-part__block-control-group">
-					<ToolbarButton onClick={ openPatternSelectionModal }>
-						{ __( 'Replace' ) }
-					</ToolbarButton>
-				</ToolbarGroup>
-			) }
-		</>
+		<ToolbarGroup className="wp-block-template-part__block-control-group">
```

Comment:
Not for this PR but how much of this toolbar control can be made generic and absorbed automatically for all blocks in "block-editor"? (just asking whether that's even possible)

Reply:
I had a poke around and I think it could be genericized in the block-editor package. 

For example the [align hook displays a block control](https://github.com/WordPress/gutenberg/blob/5797f50e839142856763b2152a11a655c70f4c6c/packages/block-editor/src/hooks/align.js#L144-L144).

As far as I can tell the `hasPatterns` check the query block uses relies only on the blocks and block-editor stores. So no core data.